File: clients/ios/Lextures/Core/Design/LexturesTheme.swift
import SwiftUI

/// Lextures by StudyDrift brand system.
///
/// Derived from the logo: a rocket lifting off an open book. Warm cream paper,
/// deep-teal ink, coral energy, amber highlights. Serif display type for an
/// editorial, scholarly feel; system sans for body copy.
enum LexturesTheme {
    // MARK: Brand anchors (from logo)
    static let brandTeal = Color(hex: 0x6EC0B1)   // logo teal — glows, tints
    static let brandCoral = Color(hex: 0xF6684B)  // logo coral — energy accent
    static let brandAmber = Color(hex: 0xF69945)  // logo amber — highlights
    static let brandCream = Color(hex: 0xF4E4C0)  // logo cream — paper tint

    // MARK: Action colors
    static let primary = Color(hex: 0x12756A)      // deep teal, contrast-safe
    static let primaryDeep = Color(hex: 0x0C4F47)  // gradient anchor
    static let primaryMuted = Color(hex: 0x135854) // links
    static let coral = brandCoral
    static let amber = brandAmber
    static let error = Color(hex: 0xDF3250)

    // MARK: Light surfaces
    static let sceneBackground = Color(hex: 0xFAF5EA)   // warm paper
    static let cardBackground = Color.white
    static let fieldBorder = Color(hex: 0xEAE0CC)
    static let textPrimary = Color(hex: 0x1F2D2A)       // teal-ink
    static let textSecondary = Color(hex: 0x64746F)

    // MARK: Dark surfaces (teal-tinted, never pure gray)
    static let sceneBackgroundDark = Color(hex: 0x111B19)
    static let cardBackgroundDark = Color(hex: 0x1B2725)
    static let fieldBorderDark = Color(hex: 0x32423E)
    static let textPrimaryDark = Color(hex: 0xF2EFE6)
    static let textSecondaryDark = Color(hex: 0x9CAEA8)

    static func sceneBackground(for scheme: ColorScheme) -> Color {
        scheme == .dark ? sceneBackgroundDark : sceneBackground
    }

    static func cardBackground(for scheme: ColorScheme) -> Color {
        scheme == .dark ? cardBackgroundDark : cardBackground
    }

    static func fieldBorder(for scheme: ColorScheme) -> Color {
        scheme == .dark ? fieldBorderDark : fieldBorder
    }

    static func textPrimary(for scheme: ColorScheme) -> Color {
        scheme == .dark ? textPrimaryDark : textPrimary
    }

    static func textSecondary(for scheme: ColorScheme) -> Color {
        scheme == .dark ? textSecondaryDark : textSecondary
    }

    /// Brighter primary for dark backgrounds.
    static func accent(for scheme: ColorScheme) -> Color {
        scheme == .dark ? brandTeal : primary
    }

    // MARK: Display typography (serif, editorial)
    static func displayFont(_ size: CGFloat, weight: Font.Weight = .semibold) -> Font {
        .system(size: size, weight: weight, design: .serif)
    }

    // MARK: Hero gradient (deep teal, dashboard greeting / course banners)
    static let heroGradient = LinearGradient(
        colors: [primaryDeep, Color(hex: 0x17897B)],
        startPoint: .topLeading,
        endPoint: .bottomTrailing
    )

    /// Deterministic course-cover gradient: every course gets a stable brand color.
    static func coverGradient(for key: String) -> LinearGradient {
        let palettes: [(Color, Color)] = [
            (Color(hex: 0x17897B), Color(hex: 0x6EC0B1)), // teal
            (Color(hex: 0xE2553A), Color(hex: 0xF69945)), // coral → amber
            (Color(hex: 0x0C4F47), Color(hex: 0x2BA391)), // deep teal
            (Color(hex: 0xD9822B), Color(hex: 0xF6B95A)), // amber
            (Color(hex: 0xC65441), Color(hex: 0xF6684B)), // coral
        ]
        let index = abs(key.unicodeScalars.reduce(0) { ($0 &* 31) &+ Int($1.value) }) % palettes.count
        let pair = palettes[index]
        return LinearGradient(colors: [pair.0, pair.1], startPoint: .topLeading, endPoint: .bottomTrailing)
    }

    /// Soft elevation shadow for floating cards.
    static func cardShadow(for scheme: ColorScheme) -> Color {
        scheme == .dark ? .clear : Color(hex: 0x3A2E18).opacity(0.08)
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255
        )
    }
}

/// Warm paper backdrop with teal + coral glows (auth, splash).
struct PublicAuthBackground: View {
    @Environment(\.colorScheme) private var colorScheme

    var body: some View {
        ZStack {
            LexturesTheme.sceneBackground(for: colorScheme)
                .ignoresSafeArea()

            RadialGradient(
                colors: [
                    LexturesTheme.brandTeal.opacity(colorScheme == .dark ? 0.12 : 0.18),
                    .clear,
                ],
                center: .topTrailing,
                startRadius: 8,
                endRadius: 460
            )
            .ignoresSafeArea()

            RadialGradient(
                colors: [
                    LexturesTheme.brandCoral.opacity(colorScheme == .dark ? 0.07 : 0.10),
                    .clear,
                ],
                center: .bottomLeading,
                startRadius: 8,
                endRadius: 420
            )
            .ignoresSafeArea()
        }
    }
}

struct AuthCard<Content: View>: View {
    @Environment(\.colorScheme) private var colorScheme
    @ViewBuilder var content: () -> Content

    var body: some View {
        content()
            .padding(28)
            .background(LexturesTheme.cardBackground(for: colorScheme))
            .clipShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: 22, style: .continuous)
                    .stroke(LexturesTheme.fieldBorder(for: colorScheme).opacity(colorScheme == .dark ? 0.9 : 0.5), lineWidth: 1)
            )
            .shadow(color: LexturesTheme.cardShadow(for: colorScheme), radius: 18, y: 8)
    }
}

struct AuthPrimaryButtonStyle: ButtonStyle {
    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.subheadline.weight(.semibold))
            .frame(maxWidth: .infinity)
            .padding(.vertical, 14)
            .foregroundStyle(.white)
            .background(
                LinearGradient(
                    colors: [LexturesTheme.primary, Color(hex: 0x17897B)],
                    startPoint: .leading,
                    endPoint: .trailing
                )
                .opacity(isEnabled ? 1 : 0.55)
            )
            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
            .shadow(
                color: LexturesTheme.primary.opacity(isEnabled ? 0.35 : 0),
                radius: configuration.isPressed ? 4 : 10,
                y: configuration.isPressed ? 2 : 5
            )
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .animation(.easeOut(duration: 0.15), value: configuration.isPressed)
    }
}

struct AuthTextField: View {
    let title: String
    @Binding var text: String
    var placeholder: String = ""
    var isSecure = false
    var keyboard: UIKeyboardType = .default
    var textContentType: UITextContentType?
    var autocapitalization: UITextAutocapitalizationType = .none

    @Environment(\.colorScheme) private var colorScheme
    @FocusState private var focused: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title)
                .font(.subheadline.weight(.medium))
                .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))

            AuthFieldRepresentable(
                text: $text,
                placeholder: placeholder,
                isSecure: isSecure,
                keyboard: keyboard,
                textContentType: textContentType,
                autocapitalization: isSecure ? .none : autocapitalization
            )
            .focused($focused)
            .padding(.horizontal, 14)
            .padding(.vertical, 12)
            .background(colorScheme == .dark ? Color(hex: 0x141F1D) : LexturesTheme.sceneBackground.opacity(0.6))
            .overlay(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .stroke(
                        focused ? LexturesTheme.primary : LexturesTheme.fieldBorder(for: colorScheme),
                        lineWidth: focused ? 2 : 1
                    )
            )
            .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
            .animation(.easeOut(duration: 0.15), value: focused)
        }
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/design/LexturesTheme.kt
package com.lextures.android.core.design

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalLayoutDirection
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp
import com.lextures.android.core.accessibility.AccessibilitySupport
import com.lextures.android.core.accessibility.LocalAccessibilityPreferences
import com.lextures.android.core.accessibility.rememberAccessibilityPreferences
import com.lextures.android.core.i18n.LocalLocalePreferences
import com.lextures.android.core.i18n.rememberLocalePreferences

/**
 * Lextures by StudyDrift brand system.
 *
 * Derived from the logo: a rocket lifting off an open book. Warm cream paper,
 * deep-teal ink, coral energy, amber highlights. Serif display type for an
 * editorial, scholarly feel; system sans for body copy.
 */
object LexturesColors {
    // Brand anchors (from logo)
    val BrandTeal = Color(0xFF6EC0B1)
    val BrandCoral = Color(0xFFF6684B)
    val BrandAmber = Color(0xFFF69945)
    val BrandCream = Color(0xFFF4E4C0)

    // Action colors
    val Primary = Color(0xFF12756A)
    val PrimaryDeep = Color(0xFF0C4F47)
    val PrimaryMuted = Color(0xFF135854)
    val Coral = BrandCoral
    val Amber = BrandAmber
    val Error = Color(0xFFDF3250)
    val StrengthStrong = Color(0xFF0C9857)

    // Light surfaces
    val SceneBackground = Color(0xFFFAF5EA)
    val CardBackground = Color.White
    val FieldBorder = Color(0xFFEAE0CC)
    val TextPrimary = Color(0xFF1F2D2A)
    val TextSecondary = Color(0xFF64746F)

    // Dark surfaces (teal-tinted, never pure gray)
    val SceneBackgroundDark = Color(0xFF111B19)
    val CardBackgroundDark = Color(0xFF1B2725)
    val FieldBorderDark = Color(0xFF32423E)
    val TextPrimaryDark = Color(0xFFF2EFE6)
    val TextSecondaryDark = Color(0xFF9CAEA8)
}

/** Serif display type — like a textbook chapter heading. */
object LexturesType {
    val DisplayFamily = FontFamily.Serif
    val DyslexiaFamily = FontFamily.SansSerif

    const val DYSLEXIA_LETTER_SPACING = 0.6f
    const val DYSLEXIA_LINE_HEIGHT_MULTIPLIER = 1.35f

    fun display(size: Int, weight: FontWeight = FontWeight.SemiBold, dyslexia: Boolean = false) = TextStyle(
        fontFamily = if (dyslexia) DyslexiaFamily else DisplayFamily,
        fontWeight = weight,
        fontSize = size.sp,
        letterSpacing = if (dyslexia) DYSLEXIA_LETTER_SPACING.sp else TextUnit.Unspecified,
        lineHeight = if (dyslexia) (size * DYSLEXIA_LINE_HEIGHT_MULTIPLIER).sp else TextUnit.Unspecified,
    )

    fun body(size: Int = 16, dyslexia: Boolean = false) = TextStyle(
        fontFamily = if (dyslexia) DyslexiaFamily else FontFamily.Default,
        fontSize = size.sp,
        letterSpacing = if (dyslexia) DYSLEXIA_LETTER_SPACING.sp else TextUnit.Unspecified,
        lineHeight = if (dyslexia) (size * DYSLEXIA_LINE_HEIGHT_MULTIPLIER).sp else TextUnit.Unspecified,
    )
}

@Composable
fun isDarkTheme(): Boolean = isSystemInDarkTheme()

@Composable
fun isHighContrastEnabled(): Boolean {
    val configuration = LocalConfiguration.current
    // fontScale >= 1.3 often pairs with accessibility settings; also honor system high contrast.
    return configuration.fontScale >= 1.3f
}

@Composable
fun sceneBackground(): Color =
    if (isDarkTheme()) LexturesColors.SceneBackgroundDark else LexturesColors.SceneBackground

@Composable
fun cardBackground(): Color =
    if (isDarkTheme()) LexturesColors.CardBackgroundDark else LexturesColors.CardBackground

@Composable
fun fieldBorder(): Color =
    if (isDarkTheme()) LexturesColors.FieldBorderDark else LexturesColors.FieldBorder

@Composable
fun textPrimary(): Color {
    val highContrast = isHighContrastEnabled()
    return if (highContrast) {
        if (isDarkTheme()) Color.White else Color(0xFF0A1210)
    } else if (isDarkTheme()) {
        LexturesColors.TextPrimaryDark
    } else {
        LexturesColors.TextPrimary
    }
}

@Composable
fun textSecondary(): Color {
    val highContrast = isHighContrastEnabled()
    return if (highContrast) {
        if (isDarkTheme()) Color(0xFFE8ECEB) else Color(0xFF3A4A46)
    } else if (isDarkTheme()) {
        LexturesColors.TextSecondaryDark
    } else {
        LexturesColors.TextSecondary
    }
}

/** Brighter primary for dark backgrounds. */
@Composable
fun accentColor(): Color {
    val highContrast = isHighContrastEnabled()
    return if (highContrast) {
        if (isDarkTheme()) Color(0xFF8FE0D0) else Color(0xFF0A5C52)
    } else if (isDarkTheme()) {
        LexturesColors.BrandTeal
    } else {
        LexturesColors.Primary
    }
}

/** WCAG AA contrast for primary text on scene background (light and dark). */
val primaryTextContrastMeetsAA: Boolean
    get() {
        val light = AccessibilitySupport.contrastRatio(
            AccessibilitySupport.ColorComponents(0x1F2D2A),
            AccessibilitySupport.ColorComponents(0xFAF5EA),
        )
        val dark = AccessibilitySupport.contrastRatio(
            AccessibilitySupport.ColorComponents(0xF2EFE6),
            AccessibilitySupport.ColorComponents(0x111B19),
        )
        return AccessibilitySupport.meetsWcagAA(light) && AccessibilitySupport.meetsWcagAA(dark)
    }

/** Hero gradient (deep teal): dashboard greeting, course banners. */
val HeroBrush: Brush = Brush.linearGradient(
    colors = listOf(LexturesColors.PrimaryDeep, Color(0xFF17897B)),
)

private val CoverPalettes = listOf(
    listOf(Color(0xFF17897B), Color(0xFF6EC0B1)), // teal
    listOf(Color(0xFFE2553A), Color(0xFFF69945)), // coral → amber
    listOf(Color(0xFF0C4F47), Color(0xFF2BA391)), // deep teal
    listOf(Color(0xFFD9822B), Color(0xFFF6B95A)), // amber
    listOf(Color(0xFFC65441), Color(0xFFF6684B)), // coral
)

/** Deterministic course-cover gradient: every course gets a stable brand color. */
fun coverBrush(key: String): Brush {
    val index = abs(key.fold(0) { acc, c -> acc * 31 + c.code }) % CoverPalettes.size
    return Brush.linearGradient(
        colors = CoverPalettes[index],
        start = Offset.Zero,
        end = Offset(420f, 420f),
    )
}

@Composable
fun rememberLocalePreferences(): LocalePreferences {
    val context = LocalContext.current
    return remember { LocalePreferences(context.applicationContext) }
}

@Composable
fun LexturesTheme(content: @Composable () -> Unit) {
    val preferences = rememberAccessibilityPreferences()
    val localePreferences = rememberLocalePreferences()
    val dark = isSystemInDarkTheme()
    val scheme = if (dark) {
        darkColorScheme(
            primary = LexturesColors.BrandTeal,
            onPrimary = LexturesColors.PrimaryDeep,
            background = LexturesColors.SceneBackgroundDark,
            surface = LexturesColors.CardBackgroundDark,
            onBackground = LexturesColors.TextPrimaryDark,
            onSurface = LexturesColors.TextPrimaryDark,
            secondary = LexturesColors.BrandCoral,
            tertiary = LexturesColors.BrandAmber,
            error = LexturesColors.Error,
        )
    } else {
        lightColorScheme(
            primary = LexturesColors.Primary,
            onPrimary = Color.White,
            background = LexturesColors.SceneBackground,
            surface = LexturesColors.CardBackground,
            onBackground = LexturesColors.TextPrimary,
            onSurface = LexturesColors.TextPrimary,
            secondary = LexturesColors.BrandCoral,
            tertiary = LexturesColors.BrandAmber,
            error = LexturesColors.Error,
        )
    }

    MaterialTheme(
        colorScheme = scheme,
        content = {
            val layoutDirection = if (localePreferences.isRTL) LayoutDirection.Rtl else LayoutDirection.Ltr
            CompositionLocalProvider(
                LocalAccessibilityPreferences provides preferences,
                LocalLocalePreferences provides localePreferences,
                LocalLayoutDirection provides layoutDirection,
            ) {
                content()
            }
        },
    )
}
